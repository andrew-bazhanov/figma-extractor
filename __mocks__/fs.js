import { fs } from 'memfs';

fs.__originalFs = jest.requireActual('fs');
module.exports = fs;
