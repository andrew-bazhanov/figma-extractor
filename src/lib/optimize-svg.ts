import fs from 'fs';
import { Config, optimize } from 'svgo';

const config: Config = {
  js2svg: { pretty: true },
  plugins: [
    'cleanupAttrs',
    'removeDoctype',
    'removeXMLProcInst',
    'removeComments',
    'removeMetadata',
    'removeTitle',
    'removeDesc',
    'removeUselessDefs',
    'removeEditorsNSData',
    'removeEmptyAttrs',
    'removeHiddenElems',
    'removeEmptyText',
    'removeEmptyContainers',
    'cleanupEnableBackground',
    'convertStyleToAttrs',
    'convertColors',
    'convertPathData',
    'convertTransform',
    'removeUnknownsAndDefaults',
    'removeNonInheritableGroupAttrs',
    'removeUselessStrokeAndFill',
    'removeUnusedNS',
    'cleanupIds',
    'cleanupNumericValues',
    'moveElemsAttrsToGroup',
    'moveGroupAttrsToElems',
    'collapseGroups',
    'removeRasterImages',
    'mergePaths',
    'convertShapeToPath',
    'sortAttrs',
    'removeDimensions',
    {
      name: 'removeAttrs',
      params: {
        attrs: ['(stroke|fill)'],
      },
    },
    {
      name: 'preset-default',
      params: {
        overrides: {
          removeViewBox: false,
        },
      },
    },
  ],
};

export const optimizeSvg = async (file: string) => {
  return new Promise((resolve, reject) => {
    fs.readFile(file, 'utf8', function (err, data) {
      if (err) {
        reject(err);

        return;
      }

      fs.writeFile(file, optimize(data, { path: file, ...config }).data, function (err) {
        if (err) {
          reject(err);

          return;
        }
        resolve(file);
      });
    });
  });
};
