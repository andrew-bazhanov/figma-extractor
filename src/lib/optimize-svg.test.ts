import fs from 'fs';
import { vol } from 'memfs';
import path from 'path';

import { optimizeSvg } from './optimize-svg';

jest.mock('fs');
jest.mock('fs/promises');

const svgDirectory = path.join(__dirname, '../utils/generate-styles/fixtures');

describe('optimizeSvg', () => {
  afterEach(() => {
    vol.reset();
  });

  it('should optimize exist svg file', () => {
    const fileName = '/tmp/svg/pin.svg';
    vol.fromJSON({
      [fileName]: (fs as any).__originalFs.readFileSync(path.join(svgDirectory, 'pin.svg'), {
        encoding: 'UTF-8',
      }),
    });

    expect(vol.toJSON()).toMatchSnapshot();

    return optimizeSvg(fileName).then(() => {
      expect(vol.toJSON()).toMatchSnapshot();
    });
  });

  it("should thrown the error because a file doesn't exist", () => {
    const fileName = '/tmp/svg/pin.svg';

    return expect(optimizeSvg(fileName)).rejects.toThrow(
      "ENOENT: no such file or directory, open '/tmp/svg/pin.svg'",
    );
  });

  it("should throw the error because a file can't be written", () => {
    const fileName = '/tmp/svg/pin.svg';

    vol.fromJSON({
      [fileName]: (fs as any).__originalFs.readFileSync(path.join(svgDirectory, 'pin.svg'), {
        encoding: 'UTF-8',
      }),
    });

    fs.writeFile = jest.fn() as any;

    (fs.writeFile as jest.Mock).mockImplementationOnce(
      (file: any, data: any, callback: (err: Error) => void) => {
        callback(new Error(`The file: '${file}' can't be written`));
      },
    );

    return expect(optimizeSvg(fileName)).rejects.toThrow(
      "The file: '/tmp/svg/pin.svg' can't be written",
    );
  });
});
