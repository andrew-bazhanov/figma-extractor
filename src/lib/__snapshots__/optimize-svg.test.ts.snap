// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`optimizeSvg should optimize exist svg file 1`] = `
{
  "/tmp/svg/pin.svg": "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="green">
    <path fill-rule="evenodd" d="m8.063 15.611.57.57c-.293.293-4.046 3.94-4.046 3.94a.75.75 0 0 1-1.061-1.06l3.995-3.991.542.541Z" clip-rule="evenodd"/>
    <path fill-rule="evenodd" d="M14.885 4.554a1.038 1.038 0 0 0-1.598.158l-1.982 2.974c-.35.525-.884.901-1.497 1.055L4.86 9.978a.144.144 0 0 0-.067.241l8.495 8.495a.144.144 0 0 0 .241-.067l1.237-4.95a2.538 2.538 0 0 1 1.055-1.495l2.974-1.983a1.038 1.038 0 0 0 .158-1.598l-4.067-4.067ZM13.9 2.762a2.538 2.538 0 0 1 2.044.731l4.068 4.068a2.538 2.538 0 0 1-.387 3.906l-2.974 1.983a1.038 1.038 0 0 0-.431.612l-1.238 4.949a1.644 1.644 0 0 1-2.757.764L8.633 16.18l-.556-.556-.556-.555-3.79-3.79a1.644 1.644 0 0 1 .764-2.757l4.95-1.238c.25-.062.468-.216.611-.43L12.04 3.88c.669-.864 1.105-1.043 1.862-1.118Z" clip-rule="evenodd"/>
</svg>
",
}
`;

exports[`optimizeSvg should optimize exist svg file 2`] = `
{
  "/tmp/svg/pin.svg": "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24">
    <path fill-rule="evenodd" d="m8.063 15.611.57.57c-.293.293-4.046 3.94-4.046 3.94a.75.75 0 0 1-1.061-1.06l3.995-3.991.542.541Z" clip-rule="evenodd"/>
    <path fill-rule="evenodd" d="M14.885 4.554a1.038 1.038 0 0 0-1.598.158l-1.982 2.974c-.35.525-.884.901-1.497 1.055L4.86 9.978a.144.144 0 0 0-.067.241l8.495 8.495a.144.144 0 0 0 .241-.067l1.237-4.95a2.538 2.538 0 0 1 1.055-1.495l2.974-1.983a1.038 1.038 0 0 0 .158-1.598l-4.067-4.067ZM13.9 2.762a2.538 2.538 0 0 1 2.044.731l4.068 4.068a2.538 2.538 0 0 1-.387 3.906l-2.974 1.983a1.038 1.038 0 0 0-.431.612l-1.238 4.949a1.644 1.644 0 0 1-2.757.764L8.633 16.18l-.556-.556-.556-.555-3.79-3.79a1.644 1.644 0 0 1 .764-2.757l4.95-1.238c.25-.062.468-.216.611-.43L12.04 3.88c.669-.864 1.105-1.043 1.862-1.118Z" clip-rule="evenodd"/>
</svg>
",
}
`;

exports[`optimizeSvg should optimize svg with a custom config 1`] = `
{
  "/tmp/svg/pin.svg": "<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" fill="green"><path fill-rule="evenodd" d="m8.063 15.611.57.57c-.293.293-4.046 3.94-4.046 3.94a.75.75 0 0 1-1.061-1.06l3.995-3.991.542.541Z" clip-rule="evenodd"/><path fill-rule="evenodd" d="M14.885 4.554a1.038 1.038 0 0 0-1.598.158l-1.982 2.974c-.35.525-.884.901-1.497 1.055L4.86 9.978a.144.144 0 0 0-.067.241l8.495 8.495a.144.144 0 0 0 .241-.067l1.237-4.95a2.538 2.538 0 0 1 1.055-1.495l2.974-1.983a1.038 1.038 0 0 0 .158-1.598l-4.067-4.067ZM13.9 2.762a2.538 2.538 0 0 1 2.044.731l4.068 4.068a2.538 2.538 0 0 1-.387 3.906l-2.974 1.983a1.038 1.038 0 0 0-.431.612l-1.238 4.949a1.644 1.644 0 0 1-2.757.764L8.633 16.18l-.556-.556-.556-.555-3.79-3.79a1.644 1.644 0 0 1 .764-2.757l4.95-1.238c.25-.062.468-.216.611-.43L12.04 3.88c.669-.864 1.105-1.043 1.862-1.118Z" clip-rule="evenodd"/></svg>",
}
`;
